### PR TITLE
Changing function signature and depending APIs to work with a paramater abstraction and fixed issue where warning diagnostics were not logged when creating target

### LINKED ISSUE #186: Warnings for unused bindings not appearing
Steps to reproduce:
1. Create queue trigger function
2. Go to integrate tab, add blob output and save
3. Go back to develop tab.
4. Drop message in queue and see log update with function execution (this step should be optional I think)

Expected result: Logs show warning indicating that output binding argument is missing
Actual result: No warning appears in logs.

### PR DESCRIPTION
**Pull request by fabiocav**
Resolves #186 

This also introduces a change to the `FunctionSignature` type, changing the public API to use an abstraction to represent parameters, which simplifies testing and other language implementations like F#.

### CODE REVIEW

File: src/WebJobs.Script/Description/DotNet/FunctionSignature.cs [hunk 1]
```diff
@@ -18,12 +18,12 @@ namespace Microsoft.Azure.WebJobs.Script.Description
...
     public sealed class FunctionSignature : IEquatable<FunctionSignature>
     {
-        private readonly ImmutableArray<IParameterSymbol> _parameters;
+        private readonly ImmutableArray<FunctionParameter> _parameters;
         private readonly bool _hasLocalTypeReference;
         private readonly string _parentTypeName;
         private readonly string _methodName;
 
-        public FunctionSignature(string parentTypeName, string methodName, ImmutableArray<IParameterSymbol> parameters, bool hasLocalTypeReference)
+        public FunctionSignature(string parentTypeName, string methodName, ImmutableArray<FunctionParameter> parameters, bool hasLocalTypeReference)
```
Comment (fabiocav) [inline src/WebJobs.Script/Description/DotNet/FunctionSignature.cs]: This constructor change and the property on https://github.com/Azure/azure-webjobs-sdk-script/pull/475/files#diff-98d2eb781a41e468e6a39a81a6cab4d4R46 drive most of the changes in this PR and are meant to simplify implementation of other managed languages not supported by Roslyn.

File: src/WebJobs.Script/Description/DotNet/FunctionSignature.cs [hunk 2]
```diff
@@ -67,7 +67,7 @@ public string MethodName
             }
         }
 
-        private static bool AreParametersEquivalent(IParameterSymbol param1, IParameterSymbol param2)
+        private static bool AreParametersEquivalent(FunctionParameter param1, FunctionParameter param2)
```
Comment (mathewc) [inline src/WebJobs.Script/Description/DotNet/FunctionSignature.cs]: You could probably move this to an Equals overload on your new FunctionParameter class

File: src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs [hunk 3]
```diff
@@ -295,7 +295,8 @@ private MethodInfo CreateFunctionTarget(CancellationToken cancellationToken)
                 ICompilation compilation = _compilationService.GetFunctionCompilation(Metadata);
                 FunctionSignature functionSignature = compilation.GetEntryPointSignature(_functionEntryPointResolver);
 
-                ValidateFunctionBindingArguments(functionSignature, throwIfFailed: true);
+                ImmutableArray<Diagnostic> bindingDiagnostics = ValidateFunctionBindingArguments(functionSignature, throwIfFailed: true);
```
Comment (mathewc) [inline src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs]: So you were already validating this, just not writing the results?

**Comment by mathewc** [comment]
:ship:

File: src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs (hunk 3 above)
Comment (fabiocav) [inline src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs]: That's correct. Validation still triggered a compilation exception in the case of errors. We were just missing the traces in diagnostics with lower levels (e.g. Warning)

File: src/WebJobs.Script/Description/DotNet/FunctionSignature.cs (hunk 2 above)
Comment (fabiocav) [inline src/WebJobs.Script/Description/DotNet/FunctionSignature.cs]: Good catch. This was existing code that I just updated but yes, it should absolutely be moved to the new `FunctionParameter` class. I'll do that before merging in.

**Pull request merged**